frame/support/src/traits/tokens/fungible/hold.rs

```diff
@@ -0,0 +1,393 @@
+	/// invariants such as any Existential Deposits needed or overflows/underflows.
+	/// If this cannot be done for some reason (e.g. because the account doesn't exist) then an
+	/// `Err` is returned.
+	// Implmentation note: This should increment the consumer refs if it moves total on hold from
+	// zero to non-zero and decrement in the opposite direction.
+	//
+	// Since this was not done in the previous logic, this will need either a migration or a
+	// state item which tracks whether the account is on the old logic or new.
+	fn set_balance_on_hold(
+		reason: &Self::Reason,
+		who: &AccountId,
+		amount: Self::Balance,
+	) -> DispatchResult;
+
+	/// Reduce the balance on hold of `who` by `amount`.
+	///
+	/// If `precision` is `Exact` and it cannot be reduced by that amount for
+	/// some reason, return `Err` and don't reduce it at all. If `precision` is `BestEffort`, then
+	/// reduce the balance of `who` by the most that is possible, up to `amount`.
+	///
+	/// In either case, if `Ok` is returned then the inner is the amount by which is was reduced.
+	fn decrease_balance_on_hold(
+		reason: &Self::Reason,
+		who: &AccountId,
+		mut amount: Self::Balance,
+		precision: Precision,
+	) -> Result<Self::Balance, DispatchError> {
+		let old_balance = Self::balance_on_hold(reason, who);
+		if let BestEffort = precision {
+			amount = amount.min(old_balance);
+		}
+		let new_balance = old_balance.checked_sub(&amount).ok_or(TokenError::FundsUnavailable)?;
+		Self::set_balance_on_hold(reason, who, new_balance)?;
+		Ok(amount)
+	}
+
+	/// Increase the balance on hold of `who` by `amount`.
+	///
+	/// If it cannot be increased by that amount for some reason, return `Err` and don't increase
+	/// it at all. If Ok, return the imbalance.
+	fn increase_balance_on_hold(
+		reason: &Self::Reason,
+		who: &AccountId,
+		amount: Self::Balance,
+		precision: Precision,
+	) -> Result<Self::Balance, DispatchError> {
+		let old_balance = Self::balance_on_hold(reason, who);
+		let new_balance = if let BestEffort = precision {
+			old_balance.saturating_add(amount)
+		} else {
+			old_balance.checked_add(&amount).ok_or(ArithmeticError::Overflow)?
+		};
+		let amount = new_balance.saturating_sub(old_balance);
+		if !amount.is_zero() {
+			Self::set_balance_on_hold(reason, who, new_balance)?;
+		}
+		Ok(amount)
+	}
+}
+
+/// Trait for mutating a fungible asset which can be placed on hold.
+pub trait Mutate<AccountId>:
+	Inspect<AccountId> + super::Unbalanced<AccountId> + Unbalanced<AccountId>
+{
+	/// Hold some funds in an account. If a hold for `reason` is already in place, then this
+	/// will increase it.
+	fn hold(reason: &Self::Reason, who: &AccountId, amount: Self::Balance) -> DispatchResult {
+		// NOTE: This doesn't change the total balance of the account so there's no need to
+		// check liquidity.
+
+		Self::ensure_can_hold(reason, who, amount)?;
+		// Should be infallible now, but we proceed softly anyway.
```

Comment:
This comment is a bit unclear. `We proceed softly` but the Fortitude is `Force` which doesn't sound soft.